### PR TITLE
add: type `TypeMap[A]` enabling `Tag[T]` -> instance `T`

### PR DESCRIPTION
**Pull request by hearnadam**
This will serve as the basis for both Layers + Envs that supports intersections in #203. Inspired by ZEnvironment, but hopefully more lightweight...

@kitlangton plans to PR the Envs with intersection support next.

### CODE REVIEW

**Comment by kitlangton** [comment]
Awesome! One thing that crossed my mind: Should this potentially be nested in some sub-package or within the Envs companion object? It could compete with `Envs` for autocomplete if it's exposed at the top-level.

File: kyo-core/shared/src/main/scala/kyo/env.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+extension [R](self: Env[R])
+
+    private def fatal(using t: Tag[?]): Nothing =
+        throw new RuntimeException(s"fatal: kyo.Env of contents [$self] missing value of type: [${t.show}].")
+
+    def get[A >: R](using Tag[A]): A =
+        getOrElse(fatal)
+
+    def getOrElse[A >: R](default: => A)(using t: Tag[A]): A =
+        val result = self.getOrElse(t, null)
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/env.scala:16]: Could `default` be passed to `getOrElse` directly?

File: kyo-core/shared/src/main/scala/kyo/env.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            default
+        else
+            result.asInstanceOf[A]
+        end if
+    end getOrElse
+
+    def add[A: Tag](a: A): Env[R & A] =
+        self.updated(Tag[A], a)
+
+    def union[R0](that: Env[R0]): Env[R & R0] =
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/env.scala:27]: What do you think about naming this method `add` as well?

File: kyo-core/shared/src/main/scala/kyo/env.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        self ++ that
+
+    def size: Int        = self.size
+    def isEmpty: Boolean = self.isEmpty
+
+    private[kyo] def tag: Intersection[?] = Intersection(self.keySet.toIndexedSeq)
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/env.scala:33]: I love how well the decision not to optimize intersection and union tags is paying off here! We should probably make the constructor private[kyo], though

File: kyo-core/shared/src/main/scala/kyo/package.scala [hunk 4]
```diff
@@ -63,7 +63,7 @@ package object kyo:
         ()
 
     private[kyo] inline def isNull[T](v: T): Boolean =
-        v.asInstanceOf[AnyRef] == null
+        v.asInstanceOf[AnyRef] eq null
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/package.scala:66]: thanks! :)

**Comment by fwbrasil** [comment]
This data structure seems quite useful! It's essentially a map with types as keys 😍 I imagine we might find other use cases for it as well.

File: kyo-core/shared/src/main/scala/kyo/env.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        val result = self.getOrElse(t, null)
+        if isNull(result) then
+            default
+        else
+            result.asInstanceOf[A]
+        end if
+    end getOrElse
+
+    def add[A: Tag](a: A): Env[R & A] =
+        self.updated(Tag[A], a)
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/env.scala:25]: Can you receive the `Tag` directly as an implicit parameter instead? I'm not sure if there'd be a noticeable performance difference here but avoiding the macro indirection should reduce the size of the generated bytecode.

File: kyo-core/shared/src/test/scala/kyoTest/envTest.scala [hunk 6]
```diff
@@ -0,0 +1,109 @@
+package kyoTest
+
+import kyo.*
+
+class envTest extends KyoTest:
```
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/envTest.scala:5]: Can you add tests showing that getting a type not present in the `Env` fails to compile?

File: kyo-core/shared/src/main/scala/kyo/env.scala (hunk 1 above)
Comment (kitlangton) [reply kyo-core/shared/src/main/scala/kyo/env.scala:16]: A is basically guaranteed to exist in the map, as long as we expose the right interface (and people don't unsafely cast). So I don't think we need getOrElse?

**Comment by fwbrasil** [comment]
> Awesome! One thing that crossed my mind: Should this potentially be nested in some sub-package or within the Envs companion object? It could compete with `Envs` for autocomplete if it's exposed at the top-level.

Good point! It feels like this is more generic than just for the environments use case. I think it could be in the `kyo` package but maybe with a more generic name.

File: kyo-core/shared/src/main/scala/kyo/env.scala (hunk 1 above)
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/env.scala:16]: I think we can remove it. Realistically, this method is only useful for internal usage. Later we can re-expose as package private.

File: kyo-core/shared/src/main/scala/kyo/env.scala (hunk 2 above)
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/env.scala:27]: The problem is that this not simply additive in the case of overlapping values. I think that may confuse users.

File: kyo-core/shared/src/main/scala/kyo/env.scala (hunk 5 above)
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/env.scala:25]: Sure, will do.

**Comment by kitlangton** [comment]
> Good point! It feels like this is more generic than just for the environments use case. I think it could be in the `kyo` package but maybe with a more generic name.

It's sort of a type-keyed map... `TypeMap`?

File: kyo-core/shared/src/test/scala/kyoTest/envTest.scala (hunk 6 above)
Comment (hearnadam) [reply kyo-core/shared/src/test/scala/kyoTest/envTest.scala]: Done.

File: kyo-core/shared/src/main/scala/kyo/typeMap.scala [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    private inline def fatal(using t: Tag[?]): Nothing =
+        throw new RuntimeException(s"fatal: kyo.TypeMap of contents [$self] missing value of type: [${t.show}].")
+
+    inline def get[A >: R](using inline t: Tag[A]): A =
+        self.getOrElse(t, fatal).asInstanceOf[A]
+
+    inline def add[A: Tag](a: A)(using inline t: Tag[A]): TypeMap[R & A] =
+        self.updated(t, a)
+
+    def union[R0](that: TypeMap[R0]): TypeMap[R & R0] =
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/typeMap.scala:18]: Could it be `inline` as well?

**Review by fwbrasil — APPROVED** [review]

File: kyo-core/shared/src/test/scala/kyoTest/typeMapTest.scala [hunk 8]
```diff
@@ -0,0 +1,116 @@
+package kyoTest
+
+import kyo.*
+
+class typeMapTest extends KyoTest:
```
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/typeMapTest.scala:5]: Can you add tests to show that it's not possible to call `get` and `add` with intersection and union type parameters?

Comment (hearnadam) [reply kyo-core/shared/src/test/scala/kyoTest/typeMapTest.scala:5]: Done.

**Comment by hearnadam** [comment]
I think we should merge this if you think it's ready @fwbrasil. We can followup to:

1. add `.prune`
2. potential: require distinct values in `.apply`?
3. make Intersection.apply package private

**Comment by fwbrasil** [comment]
Love this! Thanks folks 🙏

**Pull request merged**